File: Easy/1356. Sort Integers by The Number of 1 Bits/1356. Sort Integers by The Number of 1 Bits.py
class Solution:
    def sortByBits(self, arr: List[int]) -> List[int]:
        # Define a custom sorting comparator
        def count_ones(n):
            # Count the number of 1 bits in the binary representation of n
            count = 0
            while n > 0:
                count += n & 1
                n >>= 1
            return count

        # Sort the array using the custom comparator
        arr.sort(key=lambda x: (count_ones(x), x))

        return arr

File: Easy/1356. Sort Integers by The Number of 1 Bits/1356. Sort Integers by The Number of 1 Bits.rs
impl Solution {
    pub fn sort_by_bits(arr: Vec<i32>) -> Vec<i32> {
        // Define a custom sorting comparator
        fn count_ones(n: i32) -> i32 {
            // Count the number of 1 bits in the binary representation of n
            let mut count = 0;
            let mut n = n;
            while n > 0 {
                count += n & 1;
                n >>= 1;
            }
            count
        }

        let mut arr = arr;
        // Sort the array using counting sort
        arr.sort_by_cached_key(|&x| (count_ones(x), x));

        arr
    }
}
